File: scripts/rsc-to-csv.py
#!/usr/bin/env python3
"""
Convert MikroTik /ip hotspot user export (.rsc) to hotspot-pay import CSV.

Usage:
  python rsc-to-csv.py path/to/export.rsc
  python rsc-to-csv.py path/to/export.rsc -o data/new-vouchers.csv

Output columns: code,profile
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path


def parse_rsc(text: str) -> list[tuple[str, str]]:
    text = re.sub(r"\\\s*\n\s*", " ", text)
    rows: list[tuple[str, str]] = []
    seen: set[str] = set()

    for line in text.splitlines():
        line = line.strip()
        if not line.startswith("add ") or "limit-bytes-total" not in line:
            continue
        if "profile=" not in line:
            continue

        name_match = re.search(r"name=\s*([A-Z0-9]+)", line)
        profile_match = re.search(r"profile=([^\s]+)", line)
        if not name_match or not profile_match:
            continue

        code = name_match.group(1).upper()
        profile = profile_match.group(1)
        if code in seen:
            continue
        seen.add(code)
        rows.append((code, profile))

    rows.sort(key=lambda r: (r[1], r[0]))
    return rows


def main() -> int:
    parser = argparse.ArgumentParser(description="MikroTik hotspot user .rsc → CSV")
    parser.add_argument("rsc_file", help="Path to MikroTik export .rsc file")
    parser.add_argument(
        "-o",
        "--output",
        help="Output CSV path (default: same name as input with .csv)",
    )
    args = parser.parse_args()

    src = Path(args.rsc_file)
    if not src.is_file():
        print(f"Error: file not found: {src}", file=sys.stderr)
        return 1

    rows = parse_rsc(src.read_text(encoding="utf-8", errors="replace"))
    if not rows:
        print("Error: no hotspot users found in file.", file=sys.stderr)
        return 1

    out = Path(args.output) if args.output else src.with_suffix(".csv")
    lines = ["code,profile"] + [f"{code},{profile}" for code, profile in rows]
    out.write_text("\n".join(lines) + "\n", encoding="utf-8")

    counts: dict[str, int] = {}
    for _, profile in rows:
        counts[profile] = counts.get(profile, 0) + 1

    print(f"Wrote {len(rows)} codes to {out}")
    for profile in sorted(counts):
        print(f"  {profile}: {counts[profile]}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/rsc-to-csv.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

/**
 * Convert MikroTik /ip hotspot user export (.rsc) to hotspot-pay import CSV.
 *
 * Usage:
 *   php scripts/rsc-to-csv.php /path/to/export.rsc
 *   php scripts/rsc-to-csv.php /path/to/export.rsc data/new-vouchers.csv
 */

if ($argc < 2) {
    fwrite(STDERR, "Usage: php scripts/rsc-to-csv.php <export.rsc> [output.csv]\n");
    exit(1);
}

$src = $argv[1];
$out = $argv[2] ?? preg_replace('/\.rsc$/i', '.csv', $src);

if (! is_file($src)) {
    fwrite(STDERR, "Error: file not found: {$src}\n");
    exit(1);
}

$text = file_get_contents($src);
$text = preg_replace('/\\\s*\n\s*/', ' ', $text) ?? $text;

$rows = [];
$seen = [];

foreach (explode("\n", $text) as $line) {
    $line = trim($line);
    if (! str_starts_with($line, 'add ')) {
        continue;
    }
    if (! str_contains($line, 'limit-bytes-total') && ! str_contains($line, 'limit-uptime')) {
        continue;
    }
    if (! str_contains($line, 'profile=')) {
        continue;
    }

    if (! preg_match('/name=\s*([A-Z0-9]+)/', $line, $nameMatch)) {
        continue;
    }
    if (! preg_match('/profile=([^\s]+)/', $line, $profileMatch)) {
        continue;
    }

    $code = strtoupper($nameMatch[1]);
    $profile = $profileMatch[1];
    if (isset($seen[$code])) {
        continue;
    }
    $seen[$code] = true;
    $rows[] = [$code, $profile];
}

if ($rows === []) {
    fwrite(STDERR, "Error: no hotspot users found in file.\n");
    exit(1);
}

usort($rows, static fn ($a, $b) => $a[1] <=> $b[1] ?: $a[0] <=> $b[0]);

$lines = ['code,profile'];
foreach ($rows as [$code, $profile]) {
    $lines[] = "{$code},{$profile}";
}

file_put_contents($out, implode("\n", $lines)."\n");

$counts = [];
foreach ($rows as [, $profile]) {
    $counts[$profile] = ($counts[$profile] ?? 0) + 1;
}

echo 'Wrote '.count($rows)." codes to {$out}\n";
ksort($counts);
foreach ($counts as $profile => $count) {
    echo "  {$profile}: {$count}\n";
}
